mscore/plugin/pluginUpdater.cpp

```diff
@@ -0,0 +1,840 @@
+            }
+      return urls;
+      }
+
+static QDateTime GetLastModified(QString& url)
+      {
+      QNetworkAccessManager nmg;
+      QEventLoop loop;
+      QNetworkReply* reply = nmg.head(QNetworkRequest(url));
+      QObject::connect(reply, SIGNAL(finished()), &loop, SLOT(quit()));
+      loop.exec();
+      QVariant last_modified = reply->header(QNetworkRequest::LastModifiedHeader);
+      if (last_modified.isValid())
+            return last_modified.toDateTime();
+      else
+            return QDateTime();
+      }
+
+
+void ResourceManager::scanPluginUpdate()
+      {
+      for (int row = 0; row < pluginsTable->rowCount(); row++) {
+            QPushButton* install = static_cast<QPushButton*>(pluginsTable->indexWidget(pluginsTable->model()->index(row, 2)));
+            const QString& page_url = install->property("page_url").toString();
+            // if installed
+            if (mscore->getPluginManager()->isPackageInstalled(page_url)) {
+                  install->setEnabled(false);
+                  install->setText(tr("Pending…"));
+                  PluginWorker* worker = new PluginWorker(mscore->getPluginManager()->getPackageDescription(page_url), this);
+                  QtConcurrent::run(&workerThreads, worker, &PluginWorker::checkUpdate, install);
+                  }
+            }
+      }
+
+//---------------------------------------------------------
+//   updatePlugin
+//---------------------------------------------------------
+void ResourceManager::updatePlugin()
+      {
+      QPushButton* button = static_cast<QPushButton*>(sender());
+      const PluginPackageDescription& desc = mscore->getPluginManager()->getPackageDescription(button->property("page_url").toString());
+      button->setEnabled(false);
+      button->setText(tr("Pending…"));
+      QPushButton* uninstall = static_cast<QPushButton*>(pluginsTable->indexWidget(pluginsTable->model()->index(button->property("row").toInt(), 3)));
+      uninstall->setEnabled(false);
+      PluginWorker* worker = new PluginWorker(desc, this);
+      QtConcurrent::run(&workerThreads, worker, &PluginWorker::updateInstall, button);
+      }
+
+//---------------------------------------------------------
+//   downloadInstallPlugin
+//---------------------------------------------------------
+
+void ResourceManager::downloadInstallPlugin()
+      {
+      QPushButton* button = static_cast<QPushButton*>(sender());
+      button->setEnabled(false);
+      button->setText(tr("Pending…"));
+      PluginWorker* worker = new PluginWorker(this);
+      QtConcurrent::run(&workerThreads, worker, &PluginWorker::downloadInstall, button);
+      }
+
+void ResourceManager::uninstallPluginPackage()
+      {
+      QPushButton* button = static_cast<QPushButton*>(sender());
+      const QString& url = button->property("page_url").toString();
+      if (!mscore->getPluginManager()->uninstallPluginPackage(url)) {
+            button->setText(tr("Uninstall failed. Try again"));
```

Comment:
Better `Please try again`?

Reply:
That's better indeed, but my concern is that the column width would be not enough...

Reply:
OK, point taken